xdp-dump/xdpdump.h

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/******************************************************************************
+ * Multiple include protection
+ ******************************************************************************/
+#ifndef _XDPDUMP_H_
+#define _XDPDUMP_H_
+
+/******************************************************************************
+ * General definitions
+ ******************************************************************************/
+#define PERF_MMAP_PAGE_COUNT	16
+#define MAX_CPUS		256
+
+/******************************************************************************
+ * General used macros
+ ******************************************************************************/
+#ifndef offsetof
+#define offsetof(type, member) ((size_t) &((type *)0)->member)
+#endif
+
+#ifndef container_of
+#define container_of(ptr, type, member) ({                    \
+	const typeof( ((type *)0)->member ) *__mptr = (ptr);  \
+	(type *)( (char *)__mptr - offsetof(type,member) );})
+#endif
+
+#ifndef min
+#define min(x, y) ((x) < (y) ? (x) : (y))
+#endif
+
+#ifndef roundup
+#define roundup(x, y) (					\
+{							\
+	typeof(y) __y = y;				\
+	(((x) + (__y - 1)) / __y) * __y;		\
+}							\
+)
+#endif
+
+#ifndef __packed
+#define __packed __attribute__((packed))
+#endif
+
+/*****************************************************************************
+ * perf data structures
+ *****************************************************************************/
+#define MDF_DIRECTION_FEXIT 1
+
+struct pkt_trace_metadata {
+	__u32 ifindex;
+	__u32 rx_queue;
+	__u16 pkt_len;
+	__u16 cap_len;
+	__u16 flags;
+} __packed;
+
+#ifdef _LINUX_PERF_EVENT_H
```

Comment:
This is to exclude the definitions when included from the BPF file? Feels a bit like a hack; wouldn't it be better to just split the bits that are shared between userspace and BPF into a separate file?

Reply:
I would prefer to keep it all in one file. Is there a special define always set for BPF programs, like Linux does for some of it's include files?

Reply:
Hmm, don't think so. I guess you could match on the compiler, but not sure that's a better solution? What happens if you just leave out the ifdef (i.e., why can't these just stay defined as part of the BPF program)?

Reply:
It will error out as perf_event_header is not defined, and if I include any of the kernel includes all hell breaks loose ;)

Reply:
I see. Hmm, I still think checking for a particular header is not ideal. I couldn't find a way to check if we're using the BPF target in the LLVM documentation, but I guess you could just check if __clang__ is defined instead?

Reply:
That was supposed to be `__clang__` - not bold :)

Reply:
There is a \_\_bpf\_\_ especially for this, so used it...

Reply:
Oh, neat! That was what I was looking for, obviously, but didn't find it. So great detective work! ;)